### DISCUSSION TITLE
Does Binary Ninja analyze libc format string?

### DISCUSSION THREAD

**Discussion by pr0xy-t** [Vector35/binaryninja-api#3831] [Analysis] [answered]
I wrote the following toy code, compiled it, analyzed it with binary ninja, and noticed that the printf argument display looks like below pic.
```
#include<stdio.h>
int main()
{

	int n = 1;
	__asm__("movq	$2, %rdx; \
		movq	$3, %rcx; \
		movq	$4, %r8; \
		movq	$5, %r9;");
	printf("n = %d\n", n);
}
```
![Screenshot from 2023-01-26 13-43-16](https://user-images.githubusercontent.com/42575435/214760110-a9b38d45-8d2e-4f83-9833-2bff85b02e09.png)
I noticed that the number of arguments of printf function is incorrect.
Does this mean that binininja's heuristic analysis is wrong?
Also, **does it mean that Binary Ninja is not analyzing libc format string?**
(It seems to me that if any registers that could be arguments are used before the printf call in the main function, they are all interpreted as printf arguments. If so, is this a bug?)

---
The above example is a bit exaggerated, but I also noticed that the following toy code also results in incorrect analyzing results.
[fuga.zip](https://github.com/Vector35/binaryninja-api/files/10506462/fuga.zip)
```
#include<stdio.h>
int main()
{

	int n1 = 1;
	int n2 = n1 * 3;
	printf("n1 = %d\n", n1);
}
```

![Screenshot from 2023-01-26 14-18-47](https://user-images.githubusercontent.com/42575435/214763785-5df75f11-67f1-4074-924f-eed9f47270a9.png)

**Comment by psifertex** [comment, marked as answer]
Sorry for the delay in responding -- in short, we do ship type libraries for the default pritnf but they're not "smart" and able to adjust automatically. That said, there is a plugin called [Format String Analysis](https://github.com/nshp/binaryninja-printf) that can be used to automatically adjust the arguments of these printf family of functions to have the correct number of arguments.
